### ISSUE TITLE
Add a (structural) variable filter to the backend for efficient simulation

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8893]
The current runtime allows to select which variables to write to the simulation result file, by selecting [Simulation Setup | Output | Variable filter in OMEdit](https://www.openmodelica.org/doc/OpenModelicaUsersGuide/latest/omedit.html#output), or by providing a variableFilter input to the [simulate()](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html#simulate) API function, which overrides the XML setup with the simulation flag `-override=variableFilter` when calling the runtime simulation executable.

This option is handy to reduce the size of the simulation result files, in case of big models and/or lengthy simulations, and it can be changed at runtime without recompilation. However, the results that are filtered out are still computed by generated code, they are simply not saved to the simulation result file.

Algebraic variables are computed at each output time point, starting from parameters, time, and state values using the `functionAlgebraic()` function in the runtime. In some cases, e.g. when CPU-intensive computations like fluid properties are involved, it may happen that this function spends a lot of time computing variables that are eventually of no interest, and may be possibly even filtered out by the runtime variable filter. In this case, it would be good to avoid computing them entirely, excluding them from the output set of `functionAlgebraic()`. This would be particularly important in case of multirate solvers, otherwise calling `functionAlgebraic()` to emit all the variables with a resolution compatible with the fast variables could jeopardize the performance gains obtained by the multirate integration.

I would propose to make it possible to filter output variables _structurally_ at the backend level. For this purpose, we can introduce a compiler flag `--variableFilter=<regexp>`, to filter the variables that are really interesting as outputs for the end user. The backend will figure out to compute them efficiently and generate `functionAlgebraic()` accordingly.

This option should then be accessible from the OMEdit Simulation Setup | Translation Flags menu, as an input field Compile-Time Variable Filter, defaulting to `'.*'`, with the same logic as the runtime variable filter.

@sjoelund, @mahge, if you can help @kabdelhak and @phannebohm getting this functionality in MetaModelica, then implementing it in the backend would be a no-brainer.

@adeas31 can then add the OMEdit GUI and I can take care of documentation.

**Comment by phannebohm** [comment]
> I would propose to make it possible to filter output variables _structurally_ at the backend level. For this purpose, we can introduce a compiler flag `--variableFilter=<regexp>`, to filter the variables that are really interesting as outputs for the end user. The backend will figure out to compute them efficiently and generate `functionAlgebraic()` accordingly.

No. Let's use the `-override=variableFilter` to find the variables of interest at runtime and rewrite the whole `functionAlgebraic()` to only compute those and the algebraic variables they depend on. I have a truly marvelous concept on how to do it which this comment is too narrow to contain. But basically I propose to move the BLT-graph structure into the C runtime and do a graph traversal at each call to `functionAlgebraic()`. This would make it more flexible and I want to do that anyway for my ideas on faster event search.

**Comment by casella** [comment]
@phannebohm this was mentioned today by @kabdelhak during a meeting. Looks good because of extreme flexibility, I'm only a bit worried about performance.

BTW, @mahge mentioned that he had perform such kind of analysis for the parmodauto module, and he had to implemented from scratch in C. If this functionality gets implemented, I understand it could also be used there.

**Comment by casella** [comment]
@phannebohm there still could be a good reason to also have a backend variable filter: for large systems, it would eliminate a lot of code generation in case you're interested in a few outputs only, which is normally the case for very large models. On the other hand, those large models (with many repeated components or arrays) will eventually be handled by the new backend without being unrolled, so maybe the advantage of a backend variable filter will fade away in the future.

**Comment by phannebohm** [comment]
> @phannebohm this was mentioned today by @kabdelhak during a meeting. Looks good because of extreme flexibility, I'm only a bit worried about performance.

True, it will probably introduce a small overhead to build the graph and analyze it. But we can keep it to a minimum by doing the graph analysis once at the beginning of simulation and then writing the ordering into a function pointer array that can be used to evaluate `functionAlgebraic()`. At a later stage that array can be rebuilt if needed, but it won't be necessary to do the graph analysis at every time step. Does that eliminate your worries @casella?

> BTW, @mahge mentioned that he had perform such kind of analysis for the parmodauto module, and he had to implemented from scratch in C. If this functionality gets implemented, I understand it could also be used there.

Sorry I don't get it? :sweat_smile: You mean we can already use some of @mahge `s work, or he will be able to use ours?
I'll start on this now, so if you already have something @mahge please get in touch with me.

**Comment by phannebohm** [comment]
> @phannebohm there still could be a good reason to also have a backend variable filter: for large systems, it would eliminate a lot of code generation in case you're interested in a few outputs only, which is normally the case for very large models.

True, also @bernhardbachmann pointed out that it should be easy to implement this in the backend, so maybe we can do both.

> On the other hand, those large models (with many repeated components or arrays) will eventually be handled by the new backend without being unrolled, so maybe the advantage of a backend variable filter will fade away in the future.

Also true, but I suppose simulation time would still improve by a backend filter if we can throw away some states, which my approach cannot do I think, at least for the integrator. We could omit calculating `der(x)` and just get bogus results which we throw away , but `x` would still be integrated. Not so with the backend filter.

**Comment by casella** [comment]
> Sorry I don't get it? 😅 You mean we can already use some of @mahge `s work, or he will be able to use ours? 

I understood @mahge had to do something similar, not sure if he did it in C or metamodelica, for details you need to talk to him.

**Comment by phannebohm** [comment]
I opened a discussion #8973 for my idea. The filter in the backend is still something we should do here.

**Comment by casella** [comment]
This is now essential for serious performance improvement when handling large models, particularly when using multi-rate integration.

**Comment by casella** [comment]
I think what we should use here is a mechanism similar to what @phannebohm implemented for the selective RHS ODE computation, see #11432 and PR #14016. This would not be handle by the backend, but rather by the generated code, with possibility of changing the output variables on the fly at runtime.

The alternative is to do this in the backend, by explicitly setting top-level output variables and then setting [-d=evalOutputOnly](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-debug-evaloutputonly). This mechanism is already available and leads to more efficient runtime code, but is not very flexible.
